### PR TITLE
Added code for TiesPropose procedure and favourite neighbour

### PR DESCRIPTION
**Pull request by meghanasoni**

### CODE REVIEW

File: include/PreferenceList.h [hunk 1]
```diff
@@ -9,6 +9,8 @@
 #include <vector>
 #include "TDefs.h"
 
+enum PreferenceOrderT { cEqual=0, cBetter, cWorse };
```
Comment (rawatamit) [inline include/PreferenceList.h:12]: Indentation should be:
enum PreferenceOrderT {
  cEqual=0,
  cBetter,
  cWorse
};

File: lib/MatchingAlgorithm.cc [hunk 2]
```diff
@@ -626,10 +607,11 @@ void CriticalRSM::ties_propose(FreeListType& free_list, VertexPtr a, PreferenceL
     auto aj_data = bookkeep_data[aj];
     auto y = aj_data.level;
     auto star = aj_data.star;
-    int preference = prefers(b, a, aj);
+    auto pref_list_b = b->get_preference_list();
+    auto preference = pref_list_b.prefers(a, aj);
 
     if (a_data.level == t) {
-      if (y < t || ((y == t || (y == t && star)) && preference == 2)) {
+      if (y < t || ((y == t || (y == t && star)) && preference == cBetter)) {
```
Comment (rawatamit) [inline lib/MatchingAlgorithm.cc:614]: Complex statements should be wrapped in round braces, like (y==t) and (preference == cBetter)

File: lib/MatchingAlgorithm.cc [hunk 3]
```diff
@@ -639,7 +621,7 @@ void CriticalRSM::ties_propose(FreeListType& free_list, VertexPtr a, PreferenceL
     }
 
     if (a_data.level == t && a_data.star) {
-      if (y < t || (y == t && (preference == 1 || preference == 2)) || (y == t && star && preference == 2)) {
+      if (y < t || (y == t && (preference == cEqual|| preference == cBetter)) || (y == t && star && preference == cBetter)) {
```
Comment (rawatamit) [inline lib/MatchingAlgorithm.cc:624]: Wrap complex statements in round braces, and consider defining booleans to simplify the if.
For e.g.
bool isBetter = (preference == cBetter);
bool isBetterOrEqual = (isBetter || (preference == cEqual));

File: lib/PreferenceList.cc [hunk 4]
```diff
@@ -82,6 +82,24 @@ bool PreferenceList::is_tied(RankType rank) const {
     return (ties_.find(rank) != ties_.end());
 }
 
+// Does b prefer a1 over a2?
+PreferenceOrderT PreferenceList::prefers(VertexPtr a1, VertexPtr a2) { 
+  // Return 0 if rank of a1 = rank of a2
```
Comment (rawatamit) [inline lib/PreferenceList.cc:87]: These comments need to be updated.

**Review by rawatamit — CHANGES_REQUESTED** [review]

File: test/TestPrefers.cc [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    std::shared_ptr<Vertex> b2 = std::make_shared<Vertex>("b2");
+    std::shared_ptr<Vertex> b3 = std::make_shared<Vertex>("b3");
+
+    PreferenceList a1_pref_list;
+    a1->get_preference_list().emplace_back(b1);
+    a1->get_preference_list().set_tie(1, b2);
+    a1->get_preference_list().set_tie(1, b3);
+
+    auto M = Matching();
+    auto G = read_graph(get_filepath(get_resources_dir(), "/weak_stable.txt")); // required only for critical RSM constructor, doesn't affect output of the function
```
Comment (rawatamit) [inline test/TestPrefers.cc]: Put this comment before the code in its own line.
Its not completely necessary, but wrap the lines to 80 column width.

File: test/TestFavouriteNeighbour.cc [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+TEST_CASE("FavouriteNeighbour tied preference list", "[matching_critical_rsm]") {
+    std::shared_ptr<Vertex> a1 = std::make_shared<Vertex>("a1");
+    std::shared_ptr<Vertex> b1 = std::make_shared<Vertex>("b1");
+    std::shared_ptr<Vertex> b2 = std::make_shared<Vertex>("b2");
+    std::shared_ptr<Vertex> b3 = std::make_shared<Vertex>("b3");
+    std::shared_ptr<Vertex> b4 = std::make_shared<Vertex>("b4");
+    std::shared_ptr<Vertex> dummy = std::make_shared<Vertex>("d"); // dummy partner for all matched vertices
+
+    auto G = read_graph(get_filepath(get_resources_dir(), "/weak_stable.txt")); // required only for critical RSM constructor, doesn't affect output of the function
```
Comment (rawatamit) [inline test/TestFavouriteNeighbour.cc:40]: Comment should go in its own line.

File: test/TestFavouriteNeighbour.cc [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    }
+}
+
+TEST_CASE("FavouriteNeighbour tied preference list", "[matching_critical_rsm]") {
+    std::shared_ptr<Vertex> a1 = std::make_shared<Vertex>("a1");
+    std::shared_ptr<Vertex> b1 = std::make_shared<Vertex>("b1");
+    std::shared_ptr<Vertex> b2 = std::make_shared<Vertex>("b2");
+    std::shared_ptr<Vertex> b3 = std::make_shared<Vertex>("b3");
+    std::shared_ptr<Vertex> b4 = std::make_shared<Vertex>("b4");
+    std::shared_ptr<Vertex> dummy = std::make_shared<Vertex>("d"); // dummy partner for all matched vertices
```
Comment (rawatamit) [inline test/TestFavouriteNeighbour.cc:38]: Comment on own line.

File: test/TestFavouriteNeighbour.cc [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    PreferenceList a1_pref_list;
+    a1_pref_list.emplace_back(b1);
+    a1_pref_list.emplace_back(b2);
+    a1_pref_list.emplace_back(b3);
+
+    VertexBookkeeping a1_data = VertexBookkeeping(0, 3, 0, 0);
+
+    auto M = Matching();
+    auto G = read_graph(get_filepath(get_resources_dir(), "/weak_stable.txt")); // required only for critical RSM constructor, doesn't affect output of the function
```
Comment (rawatamit) [inline test/TestFavouriteNeighbour.cc:22]: Comment before the code line.

**Review by rawatamit — CHANGES_REQUESTED** [review]

File: include/VertexBookkeeping.h [hunk 9]
```diff
@@ -26,6 +26,11 @@ struct VertexBookkeeping {
     // Residual capacity of this vertex
     int residual;
 
+    // map to store the marked vertices
+    std::map<VertexPtr, bool> marked;
+
+    bool star;
```
Comment (rawatamit) [inline include/VertexBookkeeping.h:32]: This variable should be documented with the use, and also the place where it is updated.

**Review by rawatamit — CHANGES_REQUESTED** [review]

**Review by rawatamit — COMMENTED** [review]
What does this commit merge?

**Comment by meghanasoni** [comment]
It merges the code for TiesPropose and CriticalPropose methods in the criticalRSM class and NProposingTiesMatching to compute weak stable matching with ties(which is basically the output when there are no critical vertices). I will update the criticalRSM method by using all these functions after they are merged

**Review by rawatamit — APPROVED** [review]

**Pull request merged**